File: plugins/remote/wastool/bin/was-update.pl
#!/usr/bin/perl
use strict;
use FindBin;
use lib "$FindBin::Bin";

use CommonConfig;
use Cwd 'abs_path';
use File::Basename;
use File::Copy;
use File::Path;
use Cwd;
use POSIX qw(uname);
use Utils;

my $rc = 0;

my @uname    = uname();
my $ostype   = $uname[0];
my $shellExt = 'sh';
my $nullDev  = '/dev/null';

if ( $ostype =~ /Windows/i ) {
    $ostype   = 'windows';
    $shellExt = 'bat';
    $nullDev  = 'NUL';
}
else {
    $ostype   = 'unix';
    $shellExt = 'sh';
    $nullDev  = '/dev/null';
}

my $homePath = $FindBin::Bin;
$homePath = abs_path("$homePath/..");
$ENV{LANG} = 'utf-8';

if ( scalar(@ARGV) != 2 ) {
    my $progName = $FindBin::Script;
    print("ERROR: Use as $progName config-name instance-name\n");
    exit(1);
}

chdir($homePath);
my $mainName   = $ARGV[0];
my $configName = $mainName;
my $insName    = $ARGV[1];
my $insPrefix  = $insName;
$insPrefix =~ s/\d*$//;

my $sectionConfig;
my $config = CommonConfig->new( "$homePath/conf", "wastool.ini" );
$sectionConfig = $config->getConfig("$configName.$insName");
my $confName = "$configName.$insName";
if ( not defined($sectionConfig) ) {
    $sectionConfig = $config->getConfig("$configName.$insPrefix");
    $confName      = "$configName.$insPrefix";
}
if ( not defined($sectionConfig) ) {
    $confName      = $configName;
    $sectionConfig = $config->getConfig("$configName");
}

my $dmgrprofile = $sectionConfig->{"dmgr_profile"};
my $standalone  = $sectionConfig->{"standalone"};
my $wasprofile  = $sectionConfig->{"was_profile"};
my $wasUser     = $sectionConfig->{"was_user"};
my $wasPwd      = $sectionConfig->{"was_pwd"};
my $needDeploy  = $sectionConfig->{"need_deploy"};
if ( not defined($needDeploy) or $needDeploy =~ /[1|true]/i ) {
    $needDeploy = 1;
}
else {
    $needDeploy = 0;
}

my $ihsRoot = $sectionConfig->{"ihs_docroot"};

my $startTimeout = $sectionConfig->{'start_timeout'};
if ( not defined($startTimeout) or $startTimeout eq '' ) {
    $startTimeout = 300;
}
my $timeout = int($startTimeout);

my $servername = $sectionConfig->{"servername"};
$servername =~ s/\s*//g;
my @serverNames = split( ",", $servername );

my $appName = $sectionConfig->{"appname"};
$appName =~ s/\s*//g;
my @appNames = split( ",", $appName );

my $appFile = $sectionConfig->{"appfile"};
$appFile =~ s/\s*//g;
my @appFiles = split( ",", $appFile );

if ( scalar(@appNames) != scalar(@appFiles) ) {
    print("ERROR: Config error, appfile number is not same as appname number.\n");
    exit(1);
}

my $pkgsDir = $sectionConfig->{"pkgs_dir"};
if ( not defined($pkgsDir) or $pkgsDir eq '' ) {
    $pkgsDir = "$homePath/pkgs";
}

if ( $needDeploy = 1 and -d $wasprofile ) {
    foreach my $appname (@appNames) {
        my $precheckUrl = $sectionConfig->{"$appname.precheckurl"};
        if ( defined($precheckUrl) and $precheckUrl ne '' ) {
            if ( not Utils::CheckUrlAvailable( $precheckUrl, "GET", $timeout ) ) {
                print("ERROR: Pre-check url $precheckUrl is not available in $timeout s, starting halt.\n");
                exit(1);
            }
        }
    }

    my $deployCmd = "$wasprofile/bin/wsadmin.$shellExt  -lang jython -user $wasUser -password $wasPwd -f $homePath/bin/was-autodeploy.py";
    $deployCmd = "\"$wasprofile/bin/wsadmin.$shellExt\"  -lang jython -user $wasUser -password $wasPwd -f \"$homePath/bin/was-autodeploy.py\""
        if ( $ostype eq 'windows' );
    if ( not defined($wasPwd) or $wasPwd eq '' ) {
        $deployCmd = "$wasprofile/bin/wsadmin.$shellExt  -lang jython -f $homePath/bin/was-autodeploy.py";
    }

    #$deployCmd = "$dmgrprofile/bin/wsadmin.sh  -lang jython -user $wasUser -password $wasPwd -f $homePath/bin/was-autodeploy.py" if ($standalone eq 0);

    $ENV{'TS_WASDEPLOYTOOL_HOME'} = $homePath;
    $ENV{'TS_WASCONF_NAME'}       = $confName;
    $ENV{'TS_WASINS_NAME'}        = $insName;

    my @serverLogInfos;
    foreach my $servername (@serverNames) {
        my $logFile = "$wasprofile/logs/$servername/SystemOut.log";
        my $logInfo = {};
        $logInfo->{server} = $servername;
        $logInfo->{path}   = $logFile;
        $logInfo->{pos}    = -1;
        my $fh = IO::File->new("<$logFile");
        if ( defined($fh) ) {
            $logInfo->{pos} = $fh->tell();
            $fh->close();
        }

        push( @serverLogInfos, $logInfo );
    }
    my $ret = system($deployCmd);

    if ( $ret != 0 and $standalone eq 1 ) {
        print("INFO: Autodeploy failed, maybe the server not started, restart it and try again.\n");
        my $stopSrvCmd = "$homePath/bin/was-stop.pl $mainName $insName";
        $stopSrvCmd = "\"$homePath/bin/was-stop.pl $mainName $insName\"" if ( $ostype eq 'windows' );

        #system("perl $stopSrvCmd");
        Utils::execCmd("perl $stopSrvCmd");

        foreach my $serverName (@serverNames) {
            my $startSrvCmd = "$wasprofile/bin/startServer.$shellExt $serverName -username $wasUser -password $wasPwd";
            $startSrvCmd = "\"$wasprofile/bin/startServer.$shellExt\" $serverName -username $wasUser -password $wasPwd" if ( $ostype eq 'windows' );
            if ( not defined($wasPwd) or $wasPwd eq '' ) {
                $startSrvCmd = "$wasprofile/bin/startServer.$shellExt $serverName";
                $startSrvCmd = "\"$wasprofile/bin/startServer.$shellExt\" $serverName" if ( $ostype eq 'windows' );
            }

            #system($startSrvCmd);
            Utils::execCmd($startSrvCmd);
        }

        #system($deployCmd);
        Utils::execCmd($deployCmd);

        my @serverLogInfos;
        foreach my $servername (@serverNames) {
            my $logFile = "$wasprofile/logs/$servername/SystemOut.log";
            my $logInfo = {};
            $logInfo->{server} = $servername;
            $logInfo->{path}   = $logFile;
            $logInfo->{pos}    = undef;
            my $fh = IO::File->new("<$logFile");
            if ( defined($fh) ) {
                $fh->seek( 0, 2 );
                $logInfo->{pos} = $fh->tell();
                $fh->close();
            }

            push( @serverLogInfos, $logInfo );
        }

        #system($stopSrvCmd);
        print("INFO: Deploy complete, it will take few minutes to sync the application to other nodes, please do not restart the servers immediately.\n");
        foreach my $appname (@appNames) {
            my $checkUrl = $sectionConfig->{"$appname.checkurl"};
            if ( defined($checkUrl) and $checkUrl ne '' ) {
                if ( Utils::CheckUrlAvailable( $checkUrl, "GET", 300, \@serverLogInfos ) ) {
                    print("INFO: App $appname started.\n");
                }
                else {
                    print("ERROR: App $appname start failed.\n");
                    $rc = 2;
                }
            }
        }
    }
    else {
        print("INFO: Deploy complete, it will take few minutes to sync the application to other nodes, please do not restart the servers immediately.\n");
        foreach my $appname (@appNames) {
            my $checkUrl = $sectionConfig->{"$appname.checkurl"};
            if ( defined($checkUrl) and $checkUrl ne '' ) {
                if ( Utils::CheckUrlAvailable( $checkUrl, "GET", 300, \@serverLogInfos ) ) {
                    print("INFO: App $appname started.\n");
                }
                else {
                    print("ERROR: App $appname start failed.\n");
                    $rc = 3;
                }
            }
        }
    }
}

if ( $needDeploy == 1 and defined($ihsRoot) and $ihsRoot ne '' and -d $ihsRoot ) {
    my $appNum = scalar(@appNames);
    for ( my $i = 0 ; $i < $appNum ; $i++ ) {
        my $hasExtract = 0;
        my $appName    = $appNames[$i];
        my $appFile    = $appFiles[$i];

        my $appfilePath;
        if ( -e $appFile or $appFile =~ /^[\/\\]/ ) {
            $appfilePath = $appFile;
            $appFile     = basename($appFile);
        }
        else {
            $appfilePath = "$pkgsDir/$insName/$appFile";
        }

        my $targetDir = '';
        if ( $appFile =~ /\.war$/ ) {

            my $ctxRoot = $sectionConfig->{ lc($appName) . ".contextroot" };
            if ( not defined($ctxRoot) or $ctxRoot eq '' ) {
                $ctxRoot = $sectionConfig->{"contextroot"};
            }
            my $ihsTargetDir = "$ihsRoot/$ctxRoot";
            $ihsTargetDir =~ s/\/\//\//g;
            if ( defined($ihsRoot) and $ihsRoot ne '' and $ihsTargetDir ne '/' and -d $ihsTargetDir ) {
                if ( -d $ihsTargetDir ) {

                    #remove the targetdir
                    print("INFO: Remove deploy dir $ihsTargetDir.\n");
                    rmtree($ihsTargetDir);
                }

                print("INFO: Extract package to $ihsTargetDir.\n");

                #my $extractCmd = "unzip -qo $appfilePath -d $ihsTargetDir;rm -rf $ihsTargetDir/WEB-INF";
                my $extractCmd = "unzip -qo $appfilePath -d $ihsTargetDir";
                $extractCmd = "7z x $appfilePath -o$ihsTargetDir" if ( $ostype eq 'windows' );

                #system($extractCmd);
                Utils::execCmd($extractCmd);
                rmtree("$ihsTargetDir/WEB-INF");

                $hasExtract = 1;
            }
        }
        elsif ( $appFile =~ /\.ear$/ ) {
            my $ihsTargetDir = $ihsRoot;

            if ( defined($ihsRoot) and $ihsRoot ne '' and $ihsTargetDir ne '/' and ( -d $ihsTargetDir ) ) {
                if ( -d $ihsTargetDir ) {

                    #remove the targetdir
                    print("INFO: Remove deploy dir $ihsTargetDir.\n");
                    rmtree($ihsTargetDir);
                }
                print("INFO: Extract package to $ihsTargetDir\n");

                #system("unzip -qo $appfilePath -d $ihsTargetDir");
                my $unzipCmd = Utils::getFileOPCmd( $appfilePath, $ihsTargetDir, $ostype, 'unzip' );

                #system($unzipCmd);
                Utils::execCmd($unzipCmd);

                my @warFiles = glob("$targetDir/*.war");
                foreach my $warFile (@warFiles) {

                    #system("unzip -qo $warFile -d $warFile.extract");
                    my $unzipCmd = Utils::getFileOPCmd( $warFile, "$warFile.extract", $ostype, 'unzip' );

                    #system($unzipCmd);
                    Utils::execCmd($unzipCmd);
                    unlink($warFile);
                    my $warContextRoot = $sectionConfig->{"$appName.$warFile.contextroot"};
                    $warContextRoot = '/' if ( not defined($warContextRoot) or $warContextRoot eq '' );
                    $warContextRoot =~ s/\/+//;
                    move( "$warFile.extract", "$targetDir/$warContextRoot" );
                    rmtree("$warFile/WEB-INF");
                }
            }
            $hasExtract = 1;
        }
        else {
            my $suffix = $appFile;
            print("ERROR: File type of $appFile is not supported.\n");
            $rc = 1;
        }
    }
}

if ( $needDeploy != 1 ) {
    print("INFO: $configName $insName need not deploy .\n");
    exit(0);
}

exit($rc);

File: plugins/remote/wastool/bin/python/was-update.py
#!/usr/bin/python
import ConfigParser
import os
import os.path
import sys
import re
import glob
import shutil
import utils

if len(sys.argv) < 2:
	progName = os.path.basename(__file__)
	print("ERROR:use as %s config-name instance-name\n" % progName)
	sys.exit(1)

homepath = os.path.split(os.path.realpath(__file__))[0]
homepath = os.path.realpath("%s/.." % homepath)
os.chdir(homepath)

config = ConfigParser.ConfigParser()
config.readfp(open('conf/was.ini'))

#test section name, sequence: mainname.insname -> mainname.insprefix -> mainname
mainname = sys.argv[1]
insname = None
if len(sys.argv) == 3:
	insname = sys.argv[2]

confname = mainname
if insname is not None:
	insprefix = re.sub(r'\d*$', '', insname)
	sections = config.sections()
	if '%s.%s' % (mainname, insname) in sections:
		confname = '%s.%s' % (mainname, insname)
	elif '%s.%s' % (mainname, insprefix) in sections:
		confname = '%s.%s' % (mainname, insprefix)
else:
	print("ERROR:use as was-update.py config-name instance-name\n")
	sys.exit(1)

dmgrprofile = None
try:
	dmgrprofile = config.get(confname, 'dmgr_profile')
except:
	pass
	
standalone = 0
try:
	standalone_str = config.get(confname, 'standalone')
	if standalone_str in ['True', 'true', '1']:
		standalone = 1
except:
	pass

wasprofile = config.get(confname,'was_profile')
appfile = config.get(confname,'appfile')
appname = config.get(confname,'appname')
servername = config.get(confname, 'servername')

was_user = config.get(confname,'was_user')
was_pwd = config.get(confname,'was_pwd')

need_deploy = 1
try:
	need_deploy_str = config.get(confname,'need_deploy')
	if need_deploy_str in ['False', 'false', '0']:
		need_deploy = 0
except:
	pass


ihs_dir = None
ihs_docroot = None
try:
	ihs_dir = config.get(confname,'ihs_dir')
	ihs_docroot = config.get(confname,'ihs_docroot')
except:
	pass

servername = servername.replace(' ', '')
servername = servername.replace("\t", '')
servernames = servername.split(',')

appname = appname.replace(' ', '')
appname = appname.replace("\t", '')
appnames = appname.split(',')

appfile = appfile.replace(' ', '')
appfile = appfile.replace("\t", '')
appfiles = appfile.split(',')

if need_deploy == 1:
	if wasprofile is not None and os.path.isdir(wasprofile):
		cmd = '%s/bin/wsadmin.sh  -lang jython -user %s -password %s -f %s/bin/was-autodeploy.py' % (wasprofile,was_user,was_pwd,homepath)
		#if standalone == 0:
		#	cmd = '%s/bin/wsadmin.sh  -lang jython -user %s -password %s -f %s/bin/was-autodeploy.py' % (dmgrprofile,was_user,was_pwd,homepath)
		
		os.environ['TS_WASDEPLOYTOOL_HOME'] = homepath
		os.environ['TS_WASCONF_NAME'] = confname
		os.environ['TS_WASINS_NAME'] = insname
		ret = os.system(cmd)
		if ret != 0 and standalone == 1:
			print("INFO: autodeploy failed, maybe the server not started, restart it and try again.\n")
			stopSrvCmd = '%s/bin/was-stop.py %s %s' % (homepath,mainname,insname)
			os.system(stopSrvCmd)
			
			for servername in servernames:
				startSrvCmd = '%s/bin/startServer.sh  %s' % (wasprofile,servername)
				os.system(startSrvCmd)
			
			os.system(cmd)
			#os.system(stopSrvCmd)
			
			print("INFO: Deploy complete, it will take few minutes to sync the application to other nodes, please do not restart the servers immediately.\n")
			for appname in appnames:
				checkurl = None
				try:
						checkurl = config.get(confname, '%s.checkurl' % appname)
				except:
						pass
				utils.checkUtilUrlAvaliable(checkurl)


if ihs_docroot is not None and ihs_docroot != '' and os.path.exists(ihs_docroot):
	i = 0
	for appname in appnames:
		hasExtracted = 0
		appfile = appfiles[i]
		i = i + 1
		if appfile.endswith('.war'):
			contextroot = '/'
			try:
				contextroot = config.get(confname, '%s.contextroot' % appname)
			except:
				contextroot = config.get(confname, 'contextroot')
	
			ihs_targetdir = '%s/%s' % (ihs_docroot, contextroot)
			ihs_targetdir.replace('//', '/')
			if ihs_docroot is not None and ihs_docroot != '' and ihs_targetdir != '/' and os.path.exists(ihs_targetdir):
				if os.path.exists(ihs_targetdir):
					#remove the targetdir
					cleanCmd = 'rm -rf %s' % ihs_targetdir
					print ('INFO: Remove deploy dir %s.\n' % ihs_targetdir)
					os.system(cleanCmd)
				print ('INFO: Extract package to %s.') % ihs_targetdir
				cmd = 'unzip -qo %s/pkgs/%s/%s -d %s' % (homepath, insname, appfile, ihs_targetdir)
				rmcmd = 'rm -rf %s/WEB-INF' % (ihs_targetdir)
				os.system(cmd)
				os.system(rmcmd)
				hasExtracted = 1
	
			#if hasExtracted == 0:
				#print("ERROR: WAS target dir:%s and IHS target dir:%s are not exists.\n" % (targetdir, ihs_targetdir))
				#print("ERROR: WAS app and IHS web doc must be deploy correctly before use this programe to update app.\n")
				
	
		#TODO: ear extract must to be tested.
		elif appfile.endswith('.ear'):
	
	 		ihs_targetdir = ihs_docroot
			if ihs_docroot is not None and ihs_docroot != '' and ihs_targetdir != '/' and os.path.exists(ihs_targetdir):
				print ('INFO: Extract package to %s.') % ihs_targetdir
				cmd = 'unzip -qo %s/pkgs/%s/%s -d %s' % (homepath, insname, appfile, ihs_targetdir)
				os.system(cmd)
				
				warfiles = glob.glob('%s/*.war' % targetdir)
				for warfile in warfiles:
					cmd = 'unzip -qo %s -d %s.extract' % (warfile, warfile)
					os.system(cmd)
					shutil.remove(warfile)
					
					warContextRoot = None
					try:
						warContextRoot = config.get(confname,'%s.%s.contextroot' % (appname, warfile))
					except:
						pass
					if warContextRoot is None or warContextRoot == '':
						warContextRoot = '/'
					
					warContextRoot = re.sub(r'\/+', '', warContextRoot)
					
					shutil.move('%s.extract' % warfile, '%s/%s' % (targetdir, warContextRoot))
					rmcmd = 'rm -rf %s/WEB-INF' % (warfile)
					os.system(rmcmd)
					
				hasExtracted = 1
	
			#if hasExtracted == 0:
		#		print("ERROR: WAS target dir:%s and IHS target dir:%s are not exists.\n" % (targetdir, ihs_targetdir))
		#		print("ERROR: WAS app and IHS web doc must be deploy correctly before use this programe to update app.\n")


